### PR TITLE
User solutions authenticated API

### PR DESCRIPTION
**Pull request by tleibovitz**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by v-xuto** [comment]
@tleibovitz Firstly, please contact the ARM team to review this PR. If you don't know how to contact, you can check the third of the `ARM API Review Checklist` in the [first comment](https://github.com/Azure/azure-rest-api-specs/pull/23585#issue-1671131794).

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (sjanamma) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:55]: PUT needs to be used to create resources. ARM RPC standards doesnt allow POST to create resources

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (sjanamma) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:187]: PUT must be used to update the resource.

**Comment by sjanamma** [comment]
@tleibovitz - API spec needs re-design. If it is a new RP, are you onboarding via RPaaS? I see that the provider level APIs(with no subscription, RG levels scope) are introduced. Any publisher/tenant level APIs will need explicit approval from PAS team. Please send an email to the following email ids to kick start the review.
SARBACRev, prasnal@microsoft.com

Also, I see that POST calls are being made to create resources. This is strictly prohibited. Post calls are also used to rename, get resources in some cases in the spec. Please follow the guidelines documented in the following places to learn more about RPC guidelines or reach out to ARM API Reviewers for any questions

https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html
https://armwiki.azurewebsites.net/api_contracts/guidelines/rpguidelines.html

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:187]: Can we use PATCH?

Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:187]: We only want to update the name

Comment (ihsanali01) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json]: Thanks for sharing.

**Comment by tleibovitz** [comment]
@sjanamma Just to give some context:

1. This is not a new RP, we are introducing a new set of APIs under the same RP
2. The APIs are meant to support CRUD operations on "User Solutions". A Solution contains a collection of Marketplace products. Each Azure portal user will be able to save his own solutions and change the products composition of each of them. There is no notion of RG or subscription scope here.

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:152]: >products [](http://example.com/codeflow?start=63&length=8)

missing a GET to list all products in the solution

Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:152]: >products [](http://example.com/codeflow?start=63&length=8)

missing a GET on this URI to retrieve an individual product

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:171]: >200 [](http://example.com/codeflow?start=11&length=3)

this API is missing a request and response body

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:376]: which one of these is present on the URI?  Whichever one that is should not be part of properties and be the top level "name" property instead

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:391]: collection GET responses must only have "value" and "nextLink" properties

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/examples/GetUserSolution.json
Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/examples/GetUserSolution.json:10]: >solutions [](http://example.com/codeflow?start=39&length=9)

the resource type in the spec is "Microsoft.Marketplace/mySolutions"

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (pilor) [inline specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:377]: >products [](http://example.com/codeflow?start=9&length=8)

products should either be operated on as an array inside of a solution OR have their own child resource type where they are operated on.  Right now this spec has both which is not supported

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:152]: What about the /providers/Microsoft.Marketplace/mysolutions/{solutionId} endpoint? It is the way to get the products of the solution.

Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:152]: There is no need, products are publicly available through our public products APIs. The products collection is always required at its whole.

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:376]: You mean I should remove the 'solutionId' property from SolutionProperties 'properties' field and put it at the 'description' field level?

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:377]: We want the user to perform GET/DELETE operations on products inside some solution using only the productId, but Solution response should return an extended Product schema

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json]: removed this endpoint

Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json]: removed this endpoint

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json]: removed this endpoint

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:270]: done

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json]: done

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/examples/GetUserSolution.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/examples/GetUserSolution.json]: done

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/examples/GetUserSolution.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/examples/GetUserSolution.json]: done

File: specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json
Comment (tleibovitz) [reply specification/marketplacecatalog/resource-manager/Solutions/preview/2023-03-01-preview/solutions.json:271]: done

**Comment by v-xuto** [comment]
@tleibovitz  Please fix CI checks errors (`LintDiff`), you can refer to this [comments](https://github.com/Azure/azure-rest-api-specs/pull/23585#issuecomment-1511318088) or https://aka.ms/ci-fix on how to fix it.

**Comment by tleibovitz** [comment]
> @tleibovitz Please fix CI checks errors (`LintDiff`), you can refer to this [comments](https://github.com/Azure/azure-rest-api-specs/pull/23585#issuecomment-1511318088) or https://aka.ms/ci-fix on how to fix it. Besides, does the previous version also have this error? If these two APIs are also implemented in this way, it is not recommended to fix it.

@v-xuto - There is no subscription context in this PR. The resources are scoped to individual users across subs, RGs, tenants etc.

The two comments are referring to a case where the subscription scope is meaningful. I saw other specs with similar paths (and similar comments when running validation tools) on the repo

**Review by v-xuto — APPROVED** [review]

**Comment by raych1** [comment]
@tleibovitz , let me know if this PR is ready to merge.

**Comment by tleibovitz** [comment]
> @tleibovitz , let me know if this PR is ready to merge.

@raych1 I have added a minor change that fixes the examples. PR is ready to merge from my side

**Comment by v-xuto** [comment]
@raych1 Please help merge this PR.

**Comment by tleibovitz** [comment]
@raych1 Can you please assist with closing this PR? We are on tight schedule and it is blocking us

**Comment by chunyu3** [comment]
hello @tleibovitz The `ApiReadinessCheck` failed which means that this api-version is not deployed. we have approved and labelled the `approved-okToMerge`. After the service is deployed, you can merge the PR. Thanks

**Comment by tleibovitz** [comment]
> hello @tleibovitz The `ApiReadinessCheck` failed which means that this api-version is not deployed. we have approved and labelled the `approved-okToMerge`. After the service is deployed, you can merge the PR. Thanks

You mean the manifest deployment?

**Comment by v-xuto** [comment]
> > hello @tleibovitz The `ApiReadinessCheck` failed which means that this api-version is not deployed. we have approved and labelled the `approved-okToMerge`. After the service is deployed, you can merge the PR. Thanks
> 
> You mean the manifest deployment?

@chunyu3 Please help to answer the author's confusion.

**Comment by chunyu3** [comment]
> > hello @tleibovitz The `ApiReadinessCheck` failed which means that this api-version is not deployed. we have approved and labelled the `approved-okToMerge`. After the service is deployed, you can merge the PR. Thanks
> 
> You mean the manifest deployment?

Yes.

**Comment by tleibovitz** [comment]
@chunyu3 The manifest deployment is done. Can we merge this pr now?

**Pull request merged**
